### PR TITLE
omnisharp-roslyn: fix outdated built-in msbuild for unity

### PR DESCRIPTION
**Pull request by ericdallo**
###### Motivation for this change
The `msbuild` inside `omnisharp-roslyn` is outdated, so libs like Unity doesn't work because of that.
Also the minimum mono version for omnisharp-roslyn is now `mono6`

###### Things done
* Changed `omnisharp-roslyn` to use mono6 instead of mono5
* Changed `omnisharp-roslyn` to use nix derivation of `msbuild` instead of built-in msbuild

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by corngood** [comment]
For me this fixes these errors I was getting:

```
{"Event":"log","Body":{"LogLevel":"ERROR","Name":"OmniSharp.MSBuild.ProjectLoader","Message":"The \"GetReferenceNearestTargetFrameworkTask\" task failed unexpectedly.\nSystem.MissingMethodException: Method not found: NuGet.Frameworks.NuGetFramework NuGet.Frameworks.NuGetFramework.ParseComponents(string,string)\n  at NuGet.Build.Tasks.GetReferenceNearestTargetFrameworkTask.Execute () [0x00142] in <a45a6ef95be14d1a9ae8712d595471a7>:0 \n  at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute () [0x00029] in <4322d44f99c84aab9075efd37869f964>:0 \n  at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask (Microsoft.Build.BackEnd.ITaskExecutionHost taskExecutionHost, Microsoft.Build.BackEnd.Logging.TaskLoggingContext taskLoggingContext, Microsoft.Build.BackEnd.TaskHost taskHost, Microsoft.Build.BackEnd.ItemBucket bucket, Microsoft.Build.BackEnd.TaskExecutionMode howToExecuteTask) [0x002fc] in <4322d44f99c84aab9075efd37869f964>:0 "},"Seq":28,"Type":"event"}
```

My project loads properly now, but I still get:

```
{"Event":"log","Body":{"LogLevel":"ERROR","Name":"OmniSharp.MSBuild.ProjectLoadListener","Message":"Unexpected exception got thrown from project load listener: System.NullReferenceException: Object reference not set to an instance of an object\n  at OmniSharp.MSBuild.ProjectLoadListener.GetSdkVersion (OmniSharp.MSBuild.Notification.ProjectLoadedEventArgs args) [0x0000b] in <7e6cb753aca94549b693616e0e7d1016>:0 \n  at OmniSharp.MSBuild.ProjectLoadListener.ProjectLoaded (OmniSharp.MSBuild.Notification.ProjectLoadedEventArgs args) [0x0001a] in <7e6cb753aca94549b693616e0e7d1016>:0 "},"Seq":137,"Type":"event"}
```

I haven't had a chance to dig into this yet.

**Comment by SuperSandro2000** [comment]
@ericdallo please fix the merge commit and address the review comments.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -28061,7 +28061,7 @@ in
     stdenv = crossLibcStdenv;
     };
 
-  omnisharp-roslyn = callPackage ../development/tools/omnisharp-roslyn { };
+  omnisharp-roslyn = callPackage ../development/tools/omnisharp-roslyn { mono = mono6; };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:28064]:
```suggestion
  omnisharp-roslyn = callPackage ../development/tools/omnisharp-roslyn { };
```

File: pkgs/development/tools/omnisharp-roslyn/default.nix [hunk 2]
```diff
@@ -1,7 +1,10 @@
 { stdenv
 , fetchurl
-, mono5
+, mono
```
Comment (SuperSandro2000) [inline pkgs/development/tools/omnisharp-roslyn/default.nix:3]:
```suggestion
, mono6
```

File: pkgs/development/tools/omnisharp-roslyn/default.nix [hunk 3]
```diff
@@ -25,9 +28,14 @@ stdenv.mkDerivation rec {
   installPhase = ''
     mkdir -p $out/bin
     cd ..
-		cp -r src $out/
-    ls -al $out/src
-    makeWrapper ${mono5}/bin/mono $out/bin/omnisharp \
+    cp -r src $out/
+    rm -rf $out/src/.msbuild
+    cp -r ${msbuild}/lib/mono/msbuild $out/src/.msbuild
```
Comment (SuperSandro2000) [inline pkgs/development/tools/omnisharp-roslyn/default.nix:33]: Can we use ``ln -s`` here?

**Comment by SuperSandro2000** [comment]
@ericdallo please fix the merge commit (again...).

File: pkgs/development/tools/omnisharp-roslyn/default.nix (hunk 3 above)
Comment (ericdallo) [reply pkgs/development/tools/omnisharp-roslyn/default.nix:33]: I don't think so, because on the line below, we need to rename the folder from bin to BIn

**Comment by ericdallo** [comment]
Done @SuperSandro2000

**Comment by samhedin** [comment]
Is it possible for this to be merged soon? :heart:

**Comment by ericdallo** [comment]
c/c @SuperSandro2000

**Comment by ericdallo** [comment]
@GrahamcOfBorg build omnisharp-roslyn
@GrahamcOfBorg test omnisharp-roslyn

File: pkgs/development/tools/omnisharp-roslyn/default.nix [hunk 4]
```diff
@@ -26,8 +29,13 @@ stdenv.mkDerivation rec {
     mkdir -p $out/bin
     cd ..
     cp -r src $out/
-    ls -al $out/src
-    makeWrapper ${mono5}/bin/mono $out/bin/omnisharp \
+    rm -rf $out/src/.msbuild
```
Comment (SuperSandro2000) [inline pkgs/development/tools/omnisharp-roslyn/default.nix:32]:
```suggestion
    rm -r $out/src/.msbuild
```

**Pull request merged**